### ISSUE TITLE
tm_year is unitialized, STARTTLS IO error null, fallback to plain, depends on arch

### ISSUE THREAD

**Issue by purevg** [OpenSMTPD/OpenSMTPD#1227]
I did not catch this issue until now because opensmtpd downgrades to plain when STARTTLS is failing. Thus my emails were being sent unencrypted.

The issue is that `lt->tm_year` is unitialized in `UTCTIME_LENGTH` case, causing the start date of the certificate `before_tm` to be invalid.

I first noticed the issue on the release 7.3 branch, but it is the same result on latest commit on master branch (commit 13a504d050630d2e4dba923caeac7eb5cef2a7d0).

I took me time to converge, as before using gdb to track the issue down, I remarked the issue only appeared on armhf and i386 but not on amd64 and arm64. But finally I think this is mostly an variable alignment side effect that make it work on both 64bits arches I tested.

The following patch fixes the issue:


```
diff --git i/openbsd-compat/libtls/asn.c w/openbsd-compat/libtls/asn.c                                                                                                                                                                          
index 63b45c85..0e5b2dfe 100644                                                                                                                                                                                                                 
--- i/openbsd-compat/libtls/asn.c                                                                                                                                                                                                               
+++ w/openbsd-compat/libtls/asn.c                                                                                                                                                                                                               
@@ -111,7 +111,11 @@ ASN1_time_parse(const char *bytes, size_t len, struct tm *tm, int mode)                                                                                                                                                    
        if (lt == NULL) {                                                                                                                                                                                                                       
                memset(&ltm, 0, sizeof(ltm));                                                                                                                                                                                                   
                lt = &ltm;                                                                                                                                                                                                                      
-       }                                                                                                                                                                                                                                       
+       } else {                                                                                                                                                                                                                                
+               // tm_year is unitialized in UTCTIME_LENGTH case, initialize
+               // it first before it has a chance to be increased.
+               lt->tm_year = 0;
+        }
  
        /* Timezone is required and must be GMT (Zulu). */
        if (bytes[len - 1] != 'Z')
@@ -168,4 +172,4 @@ ASN1_time_parse(const char *bytes, size_t len, struct tm *tm, int mode)
        }
  
        return (type);
-}
\ No newline at end of file
+}
```

Before the patch:

![log_before](https://github.com/OpenSMTPD/OpenSMTPD/assets/95699360/515feb91-a68d-44b1-a7ee-42f49239cad7)

and in GDB:

![gdb_before](https://github.com/OpenSMTPD/OpenSMTPD/assets/95699360/a3bbaa8e-b4a9-4cc1-bcc7-711a6b05087f)

After the patch:
![log_after](https://github.com/OpenSMTPD/OpenSMTPD/assets/95699360/40bff366-0d40-4a34-9a39-4f4d29a9d6d5)
![gdb_after](https://github.com/OpenSMTPD/OpenSMTPD/assets/95699360/95dc6a53-1ec2-4adc-b5a4-cddaf2ec9ede)

Also, I think the error message "null" could be improved to help find quicker that the issue is related to certificate date. But I did not continue searching further.

**Comment by botovq** [comment]
Thanks. Your diff is correct. I think it would be preferable to fix this like we did in upstream quite a while back:
https://github.com/openbsd/src/commit/7d218e7d120d9b313565c256e287bbc019cb3316

**Comment by purevg** [comment]
Indeed, it would fix the issue as well. It is ok for me too. Do you want to post the update ?

**Comment by omar-polo** [comment]
Thanks for the report.  I've overlooked the compat file, apologize.  I'm looking to sync those functions with libressl and to have a release as soon as possible.

**Comment by purevg** [comment]
I confirm the update in ASN1_time_parse is also fixing the issue as expected. Thank you !

**Comment by omar-polo** [comment]
I've released 7.3.0p2 which includes this fix, thank you!
